### ISSUE TITLE
genBackEdgeMap fails on recursive subroutines

### ISSUE THREAD

**Issue by dorchard** [camfort/fortran-src#44]
Example:
```
program gbe
  implicit none
  contains
   recursive subroutine foo(x)
     integer :: x
     if (x == 0) then 
       x = x
     else
       call foo(x-1)
     end if
   end subroutine foo
end program gbe
```
`fortran-src gbe.f90 -S` fails with an exception `genBackEdgeMap` coming from the function of the same name- I guess we need to compute a fixpoint in this case when we currently are not?

**Comment by mrd** [comment]
This is actually a much deeper issue but the root cause is that the recursive subroutine is not called by anything but itself. Therefore none of its basic blocks appear in a traversal of the basic block graph when starting from the beginning of the program, simply because they are not connected to the beginning of the program.

Because the basic blocks don't appear in a reverse postorder traversal then the dataflow analysis function `dominators` ignores them as well. This produces an error because the function `genBackEdgeMap` assumes that the result of `dominators` has an entry for every basic block (by definition, every basic block dominates itself).

This part can be 'fixed' by adding stub entries for every basic block but that doesn't really solve the problem: what to do with disconnected sections of the graph? There are dominance relations within the subgraph that can be extracted, but what is the meaning of 'reverse postorder traversal' of nodes that are not connected to the beginning of the program?

**Comment by mrd** [comment]
fixed in 9de6d8e8f9ac6513fd9de61a393502f870c2aaad by refactoring `BBGr` to have a `bbgrEntries` field
